controllers/nodemaintenance_controller.go

```diff
@@ -412,7 +415,6 @@ func initMaintenanceStatus(ctx context.Context, nm *v1beta1.NodeMaintenance, dra
 			return err
 		}
 		nm.Status.TotalPods = len(podlist.Items)
-		err = r.Status().Update(ctx, nm)
```

Comment:
Moving the status update externally seems a bit risky to me (at list in this way).

The rest of the code must know that this function _might_ update the CR status and behave accordingly, which means that it is easy to change one side (this function, or the "external world") and forget to update the other side accordingly.


I would at least expect the function to return some info about the need to update the CR, e.g. some boolean "needsUpdate", or something like that

Reply:
>The rest of the code must know that this function might update the CR status 

Sure, but the function name (or its declaration) doesn't give false information about a possible update inside the function, it is only setting the CR status. If we do that immediately afterward (on success) then I think there is a high risk here. On failure, we will update it with a failure phase at `onReconcileError`...

>I would at least expect the function to return some info about the need to update the CR, e.g. some boolean "needsUpdate", or something like that

The function returns an error that can act as a boolean value. Only when there is no error then we should update the status. I think it is fine this way, as an additional boolean or an indication would be quite redundant.

Reply:
IIUC the function can return error NIL even if `nm.Status.Phase != ""` and not need CR update in this case

Reply:
I was aiming at removing the update at all in the other comment thread, but Or has a point there, that because of the long eviction duration the reconcile can take a while, so an immediate status update makes sense.
Carlo has a point here as well, that we do an unneeded update in case the CR was initialized already. Returning an indication if the CR was modified makes sense. And in the end just brings us to the original flow, but with better separation of concerns, and better testability for this function 😎

Reply:
SGTM thanks for all the input